pkg/controllers/allocation/scheduling/preferences.go

```diff
@@ -0,0 +1,92 @@
+)
+
+type Preferences struct {
+	cache *cache.Cache
+}
+
+func NewPreferences() *Preferences {
+	return &Preferences{
+		cache: cache.New(ExpirationTTL, CleanupInterval),
+	}
+}
+
+// Relax removes soft preferences from pods to enable scheduling if the cloud
+// provider's capacity is constrained. For example, this can be leveraged to
+// prefer a specific zone, but relax the preferences if the pod cannot be
+// scheduled to that zone. Preferences are removed iteratively until only hard
+// constraints remain. Pods relaxation is reset (forgotten) after 5 minutes.
+func (p *Preferences) Relax(ctx context.Context, pods []*v1.Pod) {
+	for _, pod := range pods {
+		obj, ok := p.cache.Get(string(pod.UID));
+		// Add to cache if we've never seen it before
+		if !ok {
+			p.cache.Set(string(pod.UID), pod, ExpirationTTL)
+			continue
+		}
+		// Attempt to relax the pod and update the cache
+		pod = obj.(*v1.Pod)
+		if relaxed := p.relax(ctx, pod); relaxed {
+			p.cache.Set(string(pod.UID), pod, ExpirationTTL)
+		}
+	}
+}
+
+func (p *Preferences) relax(ctx context.Context, pod *v1.Pod) bool {
+	for _, relaxFunc := range []func(*v1.Pod) bool{
+		func(pod *v1.Pod) bool { return p.removePreferredNodeAffinityTerm(pod) },
+		func(pod *v1.Pod) bool { return p.removeRequiredNodeAffinityTerm(pod) },
+	} {
+		if relaxed := relaxFunc(pod); relaxed {
+			return true
+		}
+	}
+	return false
+}
+
+func (p *Preferences) removePreferredNodeAffinityTerm(pod *v1.Pod) bool {
+	// Nothing to remove
+	if pod.Spec.Affinity == nil || pod.Spec.Affinity.NodeAffinity == nil || len(pod.Spec.Affinity.NodeAffinity.PreferredDuringSchedulingIgnoredDuringExecution) == 0 {
+		return false
+	}
+	affinity := pod.Spec.Affinity.NodeAffinity
+	// Remove preferred expressions from the first term if there's more than one
+	if len(affinity.PreferredDuringSchedulingIgnoredDuringExecution[0].Preference.MatchExpressions) > 1 {
+		affinity.PreferredDuringSchedulingIgnoredDuringExecution[0].Preference.MatchExpressions =
+			affinity.PreferredDuringSchedulingIgnoredDuringExecution[0].Preference.MatchExpressions[1:]
+		return true
+	}
+	// Remove the first term if there's more than one (terms are an OR semantic)
+	if len(affinity.PreferredDuringSchedulingIgnoredDuringExecution) > 1 {
+		affinity.PreferredDuringSchedulingIgnoredDuringExecution = affinity.PreferredDuringSchedulingIgnoredDuringExecution[1:]
+		return true
+	}
+	// Remove preferred affinity entirely
+	affinity.PreferredDuringSchedulingIgnoredDuringExecution = nil
+	return true
+}
+
+func (p *Preferences) removeRequiredNodeAffinityTerm(pod *v1.Pod) bool {
+	if pod.Spec.Affinity == nil || pod.Spec.Affinity.NodeAffinity == nil || len(pod.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms) == 0 {
+		return false
+	}
+	affinity := pod.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution
+	// Remove the first term if there's more than one (terms are an OR semantic)
```

Comment:
maybe add comment about why there's no expression-removal going on here (compared to remvoePreferred...?)

Reply:
This made me realize that it's semantically a bit confusing to relax partial requirements. New implementation now only relaxes terms.